Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp

```diff
@@ -5308,6 +5309,60 @@ RefPtr<StyleValue> Parser::parse_grid_track_size_list(Vector<ComponentValue> con
     return GridTrackSizeListStyleValue::create(CSS::GridTrackSizeList(track_list, line_names_list));
 }
 
+// https://www.w3.org/TR/css-grid-1/#grid-auto-flow-property
+RefPtr<GridAutoFlowStyleValue> Parser::parse_grid_auto_flow_value(Vector<ComponentValue> const& component_values)
+{
+    // [ row | column ] || dense
+    TokenStream<ComponentValue> tokens { component_values };
+    if (!tokens.has_next_token())
+        return nullptr;
+
+    auto parse_axis = [&]() -> Optional<GridAutoFlowStyleValue::Axis> {
+        auto transaction = tokens.begin_transaction();
+        auto token = tokens.next_token();
+        if (!token.is(Token::Type::Ident))
+            return {};
+        auto const& ident = token.token().ident();
+        if (ident.equals_ignoring_ascii_case("row"sv)) {
+            transaction.commit();
+            return GridAutoFlowStyleValue::Axis::Row;
+        } else if (ident.equals_ignoring_ascii_case("column"sv)) {
+            transaction.commit();
+            return GridAutoFlowStyleValue::Axis::Column;
+        }
+        return {};
+    };
+
+    auto parse_dense = [&]() -> Optional<GridAutoFlowStyleValue::Dense> {
+        auto transaction = tokens.begin_transaction();
+        auto token = tokens.next_token();
+        if (!token.is(Token::Type::Ident))
+            return {};
+        auto const& ident = token.token().ident();
+        if (ident.equals_ignoring_ascii_case("dense"sv)) {
+            transaction.commit();
+            return GridAutoFlowStyleValue::Dense::Yes;
+        }
+        return {};
+    };
+
+    if (auto axis = parse_axis(); axis.has_value()) {
+        if (auto dense = parse_dense(); dense.has_value()) {
+            return GridAutoFlowStyleValue::create(*axis, *dense);
+        } else {
+            return GridAutoFlowStyleValue::create(*axis, GridAutoFlowStyleValue::Dense::No);
```

Comment:
If the "dense" component fails to parse, e.g `grid-auto-flow: row green;`, shouldn't we reject it here?

Reply:
indeed, we should! I fixed that.